src/main/java/bio/terra/service/upgrade/Migrate.java

```diff
@@ -1,42 +1,308 @@
+            throw new MigrateException("Update deployment failed", ex);
+        }
+    }
+
+    public MigrateAction checkDeploymentState(String deploymentUid) {
+        final String lockTableSql = "LOCK TABLE migrate.deployment_v1 IN EXCLUSIVE MODE";
+        final String upsertTemplate = "INSERT INTO migrate.deployment_v1(dep_version, id, locking_pod_name)" +
+            " VALUES (1, '<id>', '<podname>')" +
+            " ON CONFLICT ON CONSTRAINT deployment_v1_pkey" +
+            " DO UPDATE SET id = '<id>', locking_pod_name = '<podname>'";
+
+        String upsertSql = new ST(upsertTemplate)
+            .add("podname", kubeService.getPodName())
+            .add("id", deploymentUid)
+            .render();
+
+        try (Connection connection = dataSource.getConnection();
+             PreparedStatement lockStatement = connection.prepareStatement(lockTableSql);
+             PreparedStatement statement = connection.prepareStatement(upsertSql)) {
+
+            startTransaction(connection);
+            lockStatement.execute();
+
+            DeploymentRow row = getDeploymentRow(connection);
+            if (row != null) {
+                if (StringUtils.equals(row.getId(), deploymentUid)) {
+                    if (row.getLockingPodName() == null) {
+                        // The deployment uid matches and no one has the row locked. We are good to go.
+                        commitTransaction(connection);
+                        return MigrateAction.NOTHING;
+                    } else {
+                        // The deployment uid matches, but the row is locked. We need to wait.
+                        commitTransaction(connection);
+                        return MigrateAction.WAIT_FOR_UNLOCK;
+                    }
+                }
+            }
+
+            // Either the row doesn't exist, or the uid is wrong
+            int rows = statement.executeUpdate();
+            if (rows != 1) {
+                throw new MigrateException("Failed to upsert and take the deployment lock; that should be impossible");
+            }
+            commitTransaction(connection);
+            return MigrateAction.MIGRATE;
+        } catch (SQLException ex) {
+            throw new MigrateException("Update deployment failed", ex);
+        }
+    }
+
+    private DeploymentRow getDeploymentRow(Connection connection) {
+        final String readSql = "SELECT id, locking_pod_name FROM migrate.deployment_v1 WHERE dep_version = 1";
+        DeploymentRow row = null;
+
+        try (PreparedStatement statement = connection.prepareStatement(readSql);
+             ResultSet rs = statement.executeQuery()) {
+
+            if (rs.next()) {
+                row = new DeploymentRow()
+                    .id(rs.getString("id"))
+                    .lockingPodName(rs.getString("locking_pod_name"));
+            }
+        } catch (PSQLException ex) {
+            if (!StringUtils.equals(ex.getSQLState(), "42P01")) {
```

Comment:
Nit: perhaps move this string to a static constant with a more descriptive name.

Reply:
Good point. I made a constant for it.